### ISSUE TITLE
New system for chaining checkers

### ISSUE THREAD

**Issue by swsnr** [flycheck/flycheck#836]
Edit: Take a look at the [corresponding project](https://github.com/flycheck/flycheck/projects/1) for status updates.

Our current system with `:next-checkers` works fine, but it has a couple of serious drawbacks:
- It's not at all easy to understand.
- It's overly verbose and repetitive. Just take a look at `:next-checkers` of our go checkers :unamused: 
- It's very hard to extend correctly. `flycheck-add-next-checker` can break subtly if the order of calls or the order of insertions is wrong.

I think our current system was a nice try, but it's more or less a failure in my opinion.  I didn't anticipate Flycheck's success and the amount of languages and syntax checkers we'd support as well as the desire of people to add their own syntax checkers.  

We need to come up with a better system that is able to figure out automatically which syntax checkers to run and which to skip, based on a few simple declarations.

What I have in mind is this:
- Add a `:maximum-level` property to syntax checker declarations (the exact name to be determined) which gives the maximum error level at which the syntax checker still runs, i.e. with `:maximum-level error` the syntax checker will run if there are warnings in the current buffer but not if there are errors.
- Add a `:priority` which is a number that denotes when a checker will run.  Higher priority syntax checkers run first, and only the first syntax checker of a given priority runs.

I think that simpler, but I'm really open to other suggestions!

**Comment by purcell** [comment]
Yes! It would be great to overhaul this aspect of flycheck. It's certainly a bit odd that individual checkers should specify which other checkers they should chain with. _How_ to check and _when_ to check seem like orthogonal concerns. For that matter, I'm not sure replacing `:next-checkers` with `:maximum-level` and `:priority` properties in checker definitions would improve matters.

Why not a separate nested chaining order list, or an optionally richer structure for `flycheck-checkers`?

**Comment by lunaryorn** [comment]
@purcell I'm not sure. A richer structure for `flycheck-checkers` would make it harder to add new syntax checkers to the proper place, I presume.  Likewise for a separate list that denotes chains.

I'd like to put as much information as possible into the syntax checker definition, to make setup as easy as possible.

**Comment by purcell** [comment]
I guess that if there were a way for users to update a checker's `:maximum-level` and `:priority` after it had been defined, then the same results would be possible.

**Comment by lunaryorn** [comment]
I'm not sure why the user should be able to update these?  They are specific to the syntax checker; it makes no sense to update these settings afterwards.

In particular, it's pointless to apply a linter to a syntactically invalid buffer: You can't reasonably run `hlint` if the buffer has a syntax error, hence `hlint` would have `:maximum-level warning`.  Why should you—as a user—want to change that?

Likewise for `:priority`: GHC must run before for `hlint` or otherwise we risk to run `hlint` on buffers with syntax errors.  Hence GHC would have a lower priority than `hlint`, and always run before `hlint`.  Again, why should you want to change that?  It's not a good idea to run `hlint` before GHC.

**Comment by lunaryorn** [comment]
I think you might have misunderstood me: `:maximum-level` and `:priority` are only to tell Flycheck in which order to pick syntax checkers automatically.  They are not for the user to define their preferred checker.

Flycheck provides `flycheck-disabled-checkers` and `flycheck-checker` for this purpose.

**Comment by purcell** [comment]
Nope, I think I understood you, but I don't think I was explaining myself very clearly: my point was that if each checker defines a `:maximum-level` and `:priority` in order for `flycheck` to choose checkers automatically, then you could also allow users to `setq` different values for those fields afterwards, thus providing a mechanism for fine-tuning the checker selection.

**Comment by lunaryorn** [comment]
@purcell I would rather not like to users to change these.  It's all to easy to make mistakes, and I don't know what this would add over `flycheck-disabled-checkers`.  You could probably change the order of syntax checker execution, but we choose the order for a reason; there should be no need for a user to change that.  In other words, if you feel the need to change these then that should be reported as a bug, in my opinion, because it means that Flycheck's pre-configured syntax checker chaining is flawed.

**Comment by mgudemann** [comment]
I think there's often a partial order between the checkers, i.e., for Haskell `ghc` and `hlint`. For these, the `:priority` would specify a required ordering, other checkers could be run in any order. When a syntax checker is defined, Flycheck could compute a new feasible ordering to use. Maybe `:priority` is more a classification of checkers into something like checker, linter etc. ?

**Comment by lunaryorn** [comment]
@mgudemann I thought about that as well, but I don't think that'd work well. 

In some cases (e.g. Go) we have multiple linters (e.g. vet, errcheck), all of which we'd like to run because each linter has a very narrow target, whereas in other cases (e.g. Javascript) we have multiple linters (e.g. eslint vs. jshint, or standard versus jscs) of which only one should be run, because the mostly cover the same grounds.

We could work out "classifications" for these, for sure, but that's hard to get right generically, without ending up with classifications that don't work for just the next syntax checker being added.

**Comment by mgudemann** [comment]
@lunaryorn I fully agree, and my experience with various checkers is much more limited than yours. How about having some `one-of`and `all-off`connectors, i.e., having sets of checkers to run in parallel and others selecting one.

Is Go the only "difficult" case?

**Comment by lunaryorn** [comment]
@mgudemann How would that look like?  Can you come up with a concrete example to demonstrate what you have in mind?

I'm not sure whether Go is the only difficult case.  It's the largest syntax checker chain in Flycheck currently, but there may be others with complex chaining or complex relationships.  I didn't really take a look yet.

**Comment by mgudemann** [comment]
I thought about something like this:
```
((:high-priority . '(all-of lang-checker1 lang-checker2))
        (:medium-priority . '(one-of (all-of lang-linter1 lang-linter2)
                                     (one-of lang-linter3 lang-linter4)
                                     lang-linter5)))
```

so that Flycheck would first run `lang-checker1` and `lang-checker2` and with `medium-priority`, i.e., if the first did not report errors, then either `lang-linter1` and `lang-linter2`in parallel, or either `lang-linter3` or `lang-linter4` or `lang-linter5`.

Defining a new checker would then require adding it to an existing list or creating a new position and then computing a feasible combination, fixed as long as a new checker is added.

No idea if this would be easier than the existing technique, but this was roughly what I had in mind.

**Comment by lunaryorn** [comment]
@mgudemann I think that this would be even more complex than the current system, and specifically it'd still require that each checker knows which other checkers exist.  I'd rather like to design a system where each definition is self-contained, i.e. doesn't need to know about other checkers to seamlessly plug into the chain.

----

What about this:

```cl
(flycheck-define-checker javascript-jshint
  …
  :kind syntax ;; Or `lint', `style` (and probably more)
  :maximum-level 'warning
  :conflicts-with (javascript-eslint))
```

- `kind` tells Flycheck when to run the checker.  `syntax` comes first, then `style`, then `lint`, etc.
- `conflicts-with` disables other checkers that shouldn't run together with this checker.
- `maximum-level` is the maximum error level in the current buffer at which this checker will still run.

Flycheck would evaluate these properties as follows:

1. Find the first existing and supported checker for the current mode in `flycheck-checkers`.
2. From this checker collect all further applicable checkers, by adding each subsequent checker to the set of candidates unless it conflicts (in either way) with a checker which is already in the set of candidates.
3. Then run all these checkers in order of their kind, skipping over any checkers whose maximum level is exceeded.

There'd be a couple of drawbacks from this system, though:

- We'd have to check all predicates upfront, or require that all checkers have a `:mode` (the latter would be a good idea anyway, imho)
- The system wouldn't be stateless anymore: We'd have to compute all candidates upfront and drag them along until the syntax check is done, because we always need the entire candidate set for conflict resolution.
- Running a syntax check would be considerably more involved because we'd have to keep track of candidates and the kind we're currently running.

The advantages would be:

- A new syntax checker would seamlessly plug into the chain of checkers, without having to modify existing checker definitions.
- The order of `flycheck-checkers` would be less relevant.
- We could run all checkers of each kind in parallel, which would probably speed things up quite a bit for large chains (such as go).

@purcell @mgudemann @cpitclaudel What do you think?

**Comment by lunaryorn** [comment]
@purcell I know, but I think that at some point we have to abstract or otherwise we'll end up with a very complex system (e.g. my initial proposal for a numeric priority).

The kind would only be used for syntax checker ordering; if a checker has multiple roles you'd add it to the one that goes first.  Rubocop for instance would go into `syntax`, although it's not a actually a syntax checker.  But it does check syntax before anything else and thus should run at the very beginning, even though it also checks style.

**Comment by tko** [comment]
If the main motivation for `:maximum-level` is to stop trying to run a (linter) checker on syntactically broken buffer because you know it's not going to work, I wonder if it would be more clearly declared as something like `:requires syntax` -- meaning if any of the `syntax` level checkers give an error this checker isn't run?

Then if you have the guarantees that all the checkers that are run will provide sensible results it seems to me that maybe `:maximum-level` should be something for the user to control the amount of noise? That is, while the checkers could produce reliable results on all levels maybe the user only wants to see syntax errors first? Or maybe syntax warnings and type errors together? Warnings on one level and only errors on next? Controlling how deep you want to see the rabbit hole at once.

To me `:kind` as a word does not suggest any kind of implied ordering, maybe something like `:tier` or `:stage` would be more precise?

And if instead of fixed symbols the scheme was something like `syntax = 100`, `type = 200` you'd leave some wiggle room to allow checkers to arrange themselves in order relative to each other in the same level (as in `:stage 99` would have priority over the "normal" syntax level checkers). Though admittedly that's probably taking it too far.


Note that I don't really know anything how flycheck works and this is just a drive by comment, so feel free to treat is as such. Though I do know naming is hard and the terminology used rather caught my eye.

**Comment by lunaryorn** [comment]
@tko  Thank you for your comment.  It's great to have an outside perspective.

I really like `:stage` as a name; it's much better than `:kind`.  Thank you for the suggestion, we'll use it.

On the other hand I don't think we should discuss numeric priorities again.  A number conveys no meaning, and it's literally impossible to decide whether to put a checker at 90 or 91… what'd be the difference between these?  We can add new kinds (or stages 😃) at any time if need be, but I think we should try hard to stay within less than ten. Anything beyond that is too much imho, it'll be very intricate and probably also fragile.

As for `:requires syntax`, I'm not sure whether it's better than `:maximum-level`.  I agree that it's likely easier to understand, but it raises new questions:  Notably what if there's no checker for the `syntax` stage? Or, a more intricate variant of this question, what if a checker actually `:requires type` (i.e. a properly type-checked buffer) but there's only a `syntax` stage checker for the buffer?

With `maximum-level` these questions are easy to answer: No checkers, no errors, no violation of `maximum-level`.  But with `:requires` I can't come up with easy answers.

Also, `:requires` would be less flexible (as it'd only allow to filter for `error` level errors), but I'm not sure whether anyone needs the added flexibility—I don't think so.

`:requires` would also need a different name in any case, because the term `require` is already taken in Emacs Lisp for loading libraries with `(require)`.  We should used it for something entirely different to avoid serious confusion among users.

**Comment by lunaryorn** [comment]
PS: There's now a [project for this ticket](https://github.com/flycheck/flycheck/projects/1).  Take a look for status updates and progress.

**Comment by Fuco1** [comment]
Here's a silly idea, but why not configure this on the `major-mode` level.  Let checkers be checkers and don't specify where and how they run.

Then just add a setting variable (buffer or mode local...) `flycheck-for-major-mode` and give it a list of the checkers.

Yes, it requires adding a checker *and* updating the variable, but it is completely transparent what and how it runs and it is very simple for the users to disable checkers they don't want (throw it off the list).  You can specify the dependency chain in one place and even denote which checkers run parallel etc.

I see it as a property of the major mode to have a checker enabled, not the property of the checker (which, granted, does only make sense in some mode, but still it is in my mind just some code where code goes in and errors come out).

**Comment by lunaryorn** [comment]
@Fuco1 I see it as the property of a checker to declare which major modes it supports—it makes no sense to try and use Clang for Emacs Lisp in any case—so we might be approaching the problem from two different angles, and will probably end up agreeing to disagree, but nonetheless I'm curious about how you imagine your proposal to look like in code—there's no such thing as mode-local variables so I'm not sure whether I can follow your ideas.

Could you perhaps outline an example?  Maybe take the largest chain that we currently have (Go that is), and show how it'd look like in your system?

I see a couple of issues in what I think you system is, but I'd like to make sure that I understood your proposal before I comment.

**Comment by Fuco1** [comment]
Re 1, the checker can declare the mode, but as mentioned before, because it doesn't know what else is there and how to properly place it in the chain, it is maybe only useful to foolproof the system.

About the "mode local" things, you can do that by adding a function to the major-mode hook which would be responsible for setting up flycheck.  You can have an universal one or one for each mode.

Then it can take the values from some global configuration (like `flycheck-checkers`) but where all the relationships are listed.  Because flycheck is rather centralized this will work well.

As for the specification, I was simply imagining a [preorder](https://en.wikipedia.org/wiki/Preorder) on the checkers, which you can declare with a matrix or just as the relation pairs or in any other way.

But that would end up a lot like what we have now, except instead of in 6 places it would be in one.  When I wrote the above I haven't had a clear idea of the fact that a) checkers can be missing, b) the next checkers can run on different error levels, but those are defined locally (that is, run linter only if parser did not emit a warning, but run second linter even if the previous linter emited errors because we want all the errors collected... thus we can't reuse its declaration from the parser where it had a `warning` next-level).

So it's probably no good :)

(my inspiration came from expand-region and smartparens but the parallel soon broke as this here is a bit more delicate problem)

**Comment by Fuco1** [comment]
Hm... let me try again :D I think what I had in mind was essentially a "Makefile".  Instead of saying what should run next, you say what has to run before.  That way adding a new checker does not require going back and adding it to "next clause" of every other checker.  All you would need to do is to say "I need parser to return no error, and this linter should run before me.".  And then the system figures out how to do it.

**Comment by lunaryorn** [comment]
We _had_ a preorder, `:next-checkers` was exactly that, but that didn't scale which is the raison d’être for the new system.  

I'm really still not sure what you have in mind?  Why don't you try and define the Go chain (or any other non-trivial one) in your system in Emacs Lisp?  Code's the clearest way to communicate specifications 😄 

I'd really love to understand what you proposed.  You're approaching the problem from quite a different perspective, and I'm sure that your system would contribute important insights to the problem.  Even if it turns out to be a dead end I'm sure it'll still help to uncover flaws and short-comings of the system proposed in #1072.

----

As for your second comment,t to some degree that's were we'll end up with the system I'm currently writing, except that we do not even have explicit dependencies.

A syntax checker simply declares the stage they'd like to run in, and putting a checker in stage `lint` is like saying “I'm fine with a `syntax` stage checker running before me but I don't care which one”.  And with `:conflicts-with` you can also say "I'm a `lint` checker and am fine with other linters, but just not this one that I know is redundant or enforces a different style".

And then, well, "the system figures out how to do it", i.e. order checkers by stages, evict conflicting checkers, inhibit checkers, etc…

**Comment by swsnr** [comment]
This issue wasn't updated in a long time, and I am no longer interested, so I'm closing this issue to clean up my issues list.
